### ISSUE TITLE
SD card access priority scheduling

### ISSUE THREAD

**Issue by trentgill** [whimsicalraps/wslash#10]
This issue is about reducing likelihood of SD card access misses. There is a big stack of buffered SD pages, filled with the predicted next piece of audio. This includes a long-buffer ahead (in the direction of play) of the playhead, a short buffer behind (in case of reversing tape direction), and small buffers at the loop points to enable sample-accurate looping & random-access jumps to them. There is a finite amount of RAM to work with, so we have to be smart about which pages are loaded in on the fly.

Some informative data points:
* writes take 3-10x as long as reads
* changing from read to write incurs a time penalty
* accessing sequential pages (in memory) is fast, while random-access is slow
* average read times are ~2ms per page
* average write times are ~8ms per page
* worst case write times (random access) can reach 150ms (single page, generally not in a row)

In order to avoid issues where a slow SD page access causes a read-miss, we should prioritize *read* operations over writes. as reads are so much faster, we should make sure to get them loaded into RAM in case of a worst-case high-latency write operation. the memory that the page is read into, and written out of, are the same! So all *writes* come from reads. 

### fill all the pages!

This suggests that we should fill *all* of the RAM pages with predicted reads, and only run writes when there are no pages left to read into. The main challenge then becomes how to best predict the movement of the playhead.

### random access jumps

Jumping at a loop point (or with a GOTO / SEEK) command, is a moment where there is high potential for SD traffic. Acting smartly we can deal with this effortlessly though:
1) Make an ordered list of the new predicted pages
2) Look at the read-request list (including any concurrent access), and cancel all requests that are not *exactly* the first page in the new predicted list. Free the buffers that were awaiting those reads. (this is particularly important if receiving a number of GOTO requests in quick succession, as re-buffering to a new random location will likely take upward of 100ms before it's back to low-pressure zone).
3) If there is an ongoing write, cancel & requeue it (so it waits until after the new read buffer is built)
4) Go through existing buffered pages, and drop any page that is not dirty & is now out of range. This will free a number of pages if there is a lot of predicted future reads based on the playhead. They won't have hit the playhead yet, so they can just be dropped from memory altogether (ready for a new page).
5) Read-request the predicted playhead list. This can be prioritized, filling up the numerous memory slots that were opened
6) Go through the existing buffered pages that *are* dirty, and check if they are no-longer predicted. if so, queue them for writing.

/////////////////////////////////////////////////////////

This will work if we keep 2-5 uncommitted pages for queued writes.

Making sure a read access is prioritized is essential for *overdub* mode.

When a page miss occurs:
* it's better to have a gap in an overdub (but retain the old audio), than to cut out the old audio and only have the overdub
* if a page hasn't finished reading, it might cause a recording to be thrown away, or at least overwrite the portion of the page pre-read-complete

So essentially, *READs are the most important thing*.

A scheduler would:
* Keep separate list of read & write requests
* Know how many free pages are left
* Always prioritize reads ( maintain 2/3 pages ahead of tape head)
* (in future) combine adjacent reads into single access
* If free-pages count drops below some limit (2?), prioritize a write
* If there's no reads to do, check for any writes that need to happen

Basically this boils down to, reads are king, writes only matter if we're out of memory to store a new read.

**Comment by trentgill** [comment]
I'm backtracking on all this prediction business.

Each page is 170.7ms (at 1x speed), and worst case write is ~170ms, so as long as we have a whole page either side of the playhead-containing-page we should never miss a read. If we have just jumped to a loop point we might only have 1 page of buffer, so don't do writes when that happens.

Basically, so long as we prioritize read access, the current prediction setup is 100% ok, at least at 1x. If 2x speed is causing dropouts, the easiest solution is just to add an extra 'leading' proximal page, and potentially an additional 'inside' loop page, especially for the destination loop point. That way we only use 1-3 more pages for standard buffering and everything else can just be queueing up write-requests while the big read sequence happens at a loop point.

**Comment by trentgill** [comment]
This is to say, the real key to this problem is rebuilding QStream with a separate read/write queue. This likely means QStream will need to be pulled into container, or at least be a user-space library (not a generic wrLib).

**Comment by trentgill** [comment]
fixed.

reads are now prioritized always.
with observation, even at 2x speed i didn't see more than 1 write page awaiting access in the buffer.

seems solid & ready for implementing random access now.
